urlcf7851e92b_Triploblastic_CSCD350_FinalProject_tgz-pJ8-MazeGame_MazeJ8/src/MazeGeneration/EnumMaze.java

```diff
@@ -0,0 +1,172 @@
+package MazeGeneration;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Random;
+
+import MazeRoomLogic.MazeEnums.RoomType;
+
+/* Richard Sipes 
+ * CSCD350
+ * Generates a 2D maze with RoomType enums: START, EXIT, DOOR, PATH. 
+ * Includes start/exit coordinates and a ArrayStack of trace coordinates.
+ */
+
+public class EnumMaze {
+	private int size;
+	private int[] start, exit;
+	private RoomType[][] maze;
+	private ArrayStack<int[]> trace = new ArrayStack();
+	
+	public EnumMaze(int num) {
+		if (num % 2 == 0) 
+			throw new IllegalArgumentException("Maze size must be an odd number.");
+		size = num;
+		do {
+			randomPath(maze = initMaze(size), start = randomEdgeLoc(size), trace);
+			maze[start[0]][start[1]] = RoomType.START;
+			exit = trace.peek();
+			maze[exit[0]][exit[1]] = RoomType.EXIT;
+		} while (start[0] == exit[0] && start[1] == exit[1]); //generate mazes until start != exit (usually on first try)
+	}
+	
+	/*getters*/
+	public RoomType[][] getMaze() {return this.maze;}
+	public int getSize() {return this.size;}
+	public int[] getStart() {return this.start;}
+	public int[] getExit() {return this.exit;}
+	public ArrayStack<int[]> getTrace() {return this.trace;}
+	
+	
+	private static void randomPath(RoomType[][] maze, int[] start, ArrayStack<int[]> trace) {
+		recursion(start[0],start[1], maze, trace);	
+	}
+	
+	/*First-depth-search recursive method modified from  http://www.migapro.com/depth-first-search/ */
+	/*advances the path 2 squares & pushes coordinates to the trace stack*/
+	/*doors randomly placed before backtracking*/
+	private static void recursion(int r, int c, RoomType[][] maze, ArrayStack<int[]> trace) { 
+		// 4 random directions
+	     Integer[] randDirs = generateRandomDirections();
+	     // Examine each direction
+	     for (int i = 0; i < randDirs.length; i++) {
+	    	 double potentialDoor = Math.random();
+	         
+	    	 switch(randDirs[i]){
+	         case 1: // Up
+	             //bounds check
+	             if (r - 2 <= 0)
+	                 continue;
+	             if (maze[r - 2][c] != RoomType.PATH) {
+	            	 int[] coord2 = {r-2,c}; int[] coord1 = {r-1,c};
+	            	 trace.push(coord1); trace.push(coord2);
+	            	 maze[r-2][c] = RoomType.PATH;
+	                 maze[r-1][c] = RoomType.PATH;
+	                 recursion(r - 2, c, maze, trace);
+	             } else if (potentialDoor < 0.1)
+	            	 maze[r-1][c] = RoomType.DOOR;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 59:66, 98:105.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.